[KOGITO-8129] Async API guide (#288)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-asyncapi-based-services.adoc

:keywords: kogito, workflow, serverless, AsyncApi, quarkiverse
// links:
:async_api_spec_url: https://www.asyncapi.com/docs/reference/specification/v2.0.0
:async_quarkiverse_url: https://github.com/quarkiverse/quarkus-asyncapi
// Referenced documentation pages.
:kafka_config: xref:eventing/consume-producing-events-with-kafka.adoc
:http_config: xref:eventing/consume-produce-events-with-knative-eventing.adoc
:event_states: xref:eventing/handling-events-on-workflows.adoc

This document describes how to trigger and publish events using an link:{async_api_spec_url}[AsyncAPI] specification file.
Underneath, {product_name} uses the link:{async_quarkiverse_url}[AsyncAPI Quarkiverse extension]. 
This extension automatically generates all the required link:{smallrye_messaging_url}[Smallrye] properties to make {product_name} work with the servers declared in the AsyncAPI file. Therefore, you do not need to add any property specific to your event broker to make AsyncAPI work. This is a huge advantage over regular {event_states}[event states], where whether using {kafka_config}[Kafka] or {knative_config}[Knative], you have to provide additional configuration 

[NOTE]
====

> AsyncAPI Quarkiverse extension only support `kafka` and `http` protocols at the moment. 

Reviewer: ```suggestion
AsyncAPI Quarkiverse extension only supports `kafka` and `http` protocols at the moment. 
```

====

== Consuming an event

Let's assume there is an AsyncAPI specification file called `asyncAPI.yaml`, located in the classpath of your project. 

This file contains a `servers` definition which `protocol` is Kafka: 

.Example of an AsyncAPI kafka server definition
[source,yaml]
----
servers:
  development:
    url: localhost:9092
    description: Development server
    protocol: kafka
    protocolVersion: '1.0.0'
----

and a subscription channel named `wait`: 

.Example of an AsyncAPI subscribing channel definition
[source,yaml]
----
channels:
  wait:
    description: A message channel
    subscribe:
      operationId: consumeWait
      summary: Get messages
      message:
        $ref: '#/components/messages/message'
----

You can define a function in your workflow, of type `asyncapi`, that points to `consumerWait` operationId:

.Example of an AsyncAPI subscribing function definition
[source,json]
----
{
   "functions": [
      {
          "name": "consumeEvent",
          "type": "asyncapi",
          "operation": "asyncAPI.yaml#consumeWait" 
       }
   ]
}
----

[NOTE]
====

> Remember that, as happens with `REST` and `RPC` types, the URI part of `operation` property supports `http`, `https` and `file` schemas. If no schema is present, it assumes the specification file is on project classpath.

Reviewer: ```suggestion
Remember that, as happens with `REST` and `RPC` types, the URI part of the `operation` property supports `http`, `https`, and `file` schemas. If no schema is present, it assumes the specification file is on the project classpath.
```

= Orchestrating the AsyncAPI services
:compat-mode!:
// Metadata:
:description: Orchestration of AsyncAPI based services
:keywords: kogito, workflow, serverless, AsyncApi, quarkiverse
// links:
:async_api_spec_url: https://www.asyncapi.com/docs/reference/specification/v2.0.0
:async_quarkiverse_url: https://github.com/quarkiverse/quarkus-asyncapi
// Referenced documentation pages.
:kafka_config: xref:eventing/consume-producing-events-with-kafka.adoc
:http_config: xref:eventing/consume-produce-events-with-knative-eventing.adoc
:event_states: xref:eventing/handling-events-on-workflows.adoc

This document describes how to trigger and publish events using an link:{async_api_spec_url}[AsyncAPI] specification file.

> Underneath, {product_name} uses link:{async_quarkiverse_url}[AsyncAPI Quarkiverse extension]. 

Reviewer 2: ```suggestion
Underneath, {product_name} uses the link:{async_quarkiverse_url}[AsyncAPI Quarkiverse extension]. 
```

> This extension automatically generates all the required link:{smallrye_messaging_url}[Smallrye] properties to make {product_name} work with the servers declared in the AsyncAPI file. Therefore, you do not need to add any property specific to your event broker to make AsyncAPI work. This is a huge advantage over regular link:{event_states}[event states], where whether using link:{kafka_config}[Kafka] or link:{knative_config}[Knative], you have to provide additional configuration 

Author: So link is optional, good to know ;)

> This extension automatically generates all the required link:{smallrye_messaging_url}[Smallrye] properties to make {product_name} work with the servers declared in the AsyncAPI file. Therefore, you do not need to add any property specific to your event broker to make AsyncAPI work. This is a huge advantage over regular link:{event_states}[event states], where whether using link:{kafka_config}[Kafka] or link:{knative_config}[Knative], you have to provide additional configuration 

Reviewer 3: `xref` is in fact the cross reference link :)

[NOTE]
====
AsyncAPI Quarkiverse extension only support `kafka` and `http` protocols at the moment. 
====

== Consuming an event

> Lets assume there is an AsyncAPI specification file called `asyncAPI.yaml`, located in the classpath of your project 

Reviewer 3: ```suggestion
Let's assume there is an AsyncAPI specification file called `asyncAPI.yaml`, located in the classpath of your project. 
```

> This files contains a `server` definition which `protocol` is Kafka 

Reviewer 3: ```suggestion
This file contains a `servers` definition which `protocol` is Kafka: 
```

.Example of an AsyncAPI kafka server definition
[source,yaml]
----
servers:
  development:
    url: localhost:9092
    description: Development server
    protocol: kafka
    protocolVersion: '1.0.0'
----

> and a subscription channel named `wait`. 

Reviewer 3: ```suggestion
and a subscription channel named `wait`: 
```

.Example of an AsynyAPI subscribing channel definition
[source,yaml]
----
channels:
  wait:
    description: A message channel
    subscribe:
      operationId: consumeWait
      summary: Get messages
      message:
        $ref: '#/components/messages/message'
----

You can define a function in your workflow, of type asyncapi, that points to `consumerWait` operationId.

.Example of an AsynyAPI subscribing function definition
[source,json]
----
{

>    "functions":[

Reviewer 3: ```suggestion
   "functions": [
```

      {
          "name": "consumeEvent",
          "type": "asyncapi",
          "operation": "asyncAPI.yaml#consumeWait" 
       }
   ]
}
----

> Once the function is defined, you can invoke it within an `operation` state 

Reviewer 3: ```suggestion
Once the function is defined, you can invoke it within an `operation` state: 
```

> .Example of an AsynyAPI subscribing function invocation

Reviewer 3: ```suggestion
.Example of an AsyncAPI subscribing function invocation
```

[source,json]
----
   {
      "name": "waitForEvent",
      "type": "operation",
      "actions": [
          { 
           "functionRef": "consumeEvent"
          }
       ]      
    }
----

When your flow reach `waitForEvent` state, it will stop till it receives a CloudEvent published over Kafka topic `wait`. 

[NOTE]
====
The URI part of `operation` property supports `http`, `https` and `file` schemas. If no schema is present, it assumes the specification file is on project classpath
====

== Publishing an event

Lets assume there is an AsyncAPI specification file called `asyncAPI.yaml`, located in the classpath of your project. 

This files contains a server definition which protocol is HTTP 

.Example of an AsyncAPI http server definition
[source,yaml]
----
servers:
  development:
    url: localhost:8080
    description: Development server
    protocol: http
    protocolVersion: '1.0.0'
----

and a publishing channel named `resume`

> .Example of an AsynyAPI publishing channel definition

Reviewer 3: ```suggestion
.Example of an AsyncAPI publishing channel definition
```

[source,yaml]
----
channels:
  resume:
    description: A message channel
    publish:
      operationId: sendResume
      summary: Get messages
      message:
        $ref: '#/components/messages/message'
----

You can define a function in your workflow, of type asyncapi, that point to `sendResume` operationId.

> .Example of an AsynyAPI publishing function definition

Reviewer 3: ```suggestion
.Example of an AsyncAPI publishing function definition
```

[source,json]
----
{

>    "functions":[

Reviewer 3: ```suggestion
   "functions": [
```

      {
          "name": "sendResume",
          "type": "asyncapi",
          "operation": "asyncAPI.yaml#sendResume" 
       }
   ]
}
----

> Once the function is defined, you can invoke it within an `operation` state 

Reviewer 3: ```suggestion
Once the function is defined, you can invoke it within an `operation` state:
```

> .Example of an AsynyAPI publishing function invocation

Reviewer 3: ```suggestion
.Example of an AsyncAPI publishing function invocation
```

[source,json]
----
   {
      "name": "sendEvent",
      "type": "operation",
      "actions": [
          { 
           "functionRef": "sendResume",
           "arguments" : {
              "name" : "Javierito"
            }
          }
       ]      
    }
----

> When your flow reach `sendEvent` state, it will send an event to endpoint `http://localhost:8080/resume`, with payload `{"name":"Javierito"}`

Reviewer 3: ```suggestion
When your flow reach the `sendEvent` state, it will send an event to the endpoint `http://localhost:8080/resume`, with the payload `{"name":"Javierito"}`
```

> == Additional resources

Reviewer 3: Nothing in this section ?

Author: I did not want to link to Kafka configuration because the point of async API is that you not need to configure. 
Also, you do not need to know about Event States if you master async api (except for start event state which is something being changed for 0.9)
Therefore, I did not know what to add here

Reviewer 3: then remove the section :)
